wkcuber/api/Dataset.py

```diff
@@ -0,0 +1,175 @@
+from abc import ABC, abstractmethod
+from os import mkdir
+from os.path import join, normpath, basename
+from pathlib import Path
+import numpy as np
+from os import path
+
+from wkcuber.api.Properties import WKProperties, TiffProperties
+from wkcuber.api.Layer import Layer, WKLayer, TiffLayer
+
+
+class AbstractDataset(ABC):
+    @abstractmethod
+    def __init__(self, properties):
+        self.layers = {}
+        self.path = Path(properties.path).parent
+        self.properties = properties
+
+        # construct self.layer
+        for layer_name in self.properties.data_layers:
+            layer = self.properties.data_layers[layer_name]
+            self.add_layer(
+                layer.name, layer.category, layer.element_class, layer.num_channels
+            )
+            for resolution in layer.wkw_resolutions:
+                self.layers[layer_name].setup_mag(resolution.mag.to_layer_name())
+
+    @classmethod
+    @abstractmethod
+    def open(cls, dataset_path):
+        pass
+
+    @classmethod
+    def create_with_properties(cls, properties):
+        # initialize object
+        dataset = cls(properties)
+        # create directories on disk and write datasource-properties.json
+        try:
+            mkdir(dataset.path)
+            dataset.properties.export_as_json()
+        except OSError:
+            raise FileExistsError("Creation of Dataset {} failed".format(dataset.path))
+
+        return dataset
+
+    @classmethod
+    @abstractmethod
+    def create(cls, dataset_path, scale):
+        pass
+
+    def downsample(self, layer, target_mag_shape, source_mag):
+        raise NotImplemented()
+
+    def get_properties(self):
+        return self.properties
+
+    def get_layer(self, layer_name) -> Layer:
+        if layer_name not in self.layers.keys():
+            raise IndexError(
+                "The layer {} is not a layer of this dataset".format(layer_name)
+            )
+        return self.layers[layer_name]
+
+    def add_layer(self, layer_name, category, dtype=np.dtype("uint8"), num_channels=1):
+        # normalize the value of dtype in case the parameter was passed as a string
+        dtype = np.dtype(dtype)
+
+        if layer_name in self.layers.keys():
+            raise IndexError(
+                "Adding layer {} failed. There is already a layer with this name".format(
+                    layer_name
+                )
+            )
+        self.layers[layer_name] = self.__create_layer__(layer_name, dtype, num_channels)
+        self.properties.add_layer(layer_name, category, dtype.name, num_channels)
+        return self.layers[layer_name]
+
+    def get_or_add_layer(
+        self, layer_name, category, dtype=np.dtype("uint8"), num_channels=1
+    ):
+        if layer_name in self.layers.keys():
+            assert self.properties.data_layers[layer_name].category == category, (
+                "Cannot get_or_add_layer: The layer %s already exists, but the dytpes do not match"
```

Comment:
```suggestion
                "Cannot get_or_add_layer: The layer %s already exists, but the categories do not match"
```

Also, I recommend using f-strings like this:

```
f"Cannot get_or_add_layer: The layer {layer_name} already exists, ...
```

Also, you could include the differing categories in that string. That's always very helpful when the assertion is thrown and you need to debug it :)